Enable Power Management on RT1064 SOC

danieldegrasse opened the pull request:
This PR enables power management for the RT1064 SOC. The SOC level power management hooks will handle placing the core into WAIT mode, as well as reducing core voltage. The RT1064 EVK board folder contains `lpm.c`, which provides board specific hooks to reduce clock frequencies and disable PLLs, which is required to ensure that the core voltage can be reduced without issues.

danieldegrasse committed:
soc: rt10xx: Add power management support

Add power management code for the RT10xx SOC. By default, the code will
simply gate the core clock and transition the SOC to WAIT mode when
sleeping, but if clocking hooks are register to reduce clock frequencies
these will be called as well, and the SOC core voltage will be lowered
for additional power savings.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

hakehuang commented:
@danieldegrasse  do you check this PR on clock changes? #41048

danieldegrasse commented:
@hakehuang This PR is built with reference to #41048. I'm using the same `clock_config.c` file that #41048 uses for the RT1064 clock config. Once #41048 merges, I'll rework this PR to use the `clock_config.c` file that upstream supplies. Essentially, I designed this PR with the expectation that it would depend on #41048, and only included the subset of changes from #41048 that were required to make this PR testable and passing CI.

hakehuang commented:
@danieldegrasse , this PR will make every application into low power mode, and makes the next flashing fails, this will be a big issue in CI testing. I have to switch to internal boot mode erase the nor flash and re-programing. can we write dedicated test application to enable pm(in dts overlay), and wakeup system in run mode after testing in this application, and do not change the power mode for other applications

danieldegrasse commented:
@hakehuang `CONFIG_PM` should still not be selected by default, so the debugger should only have issues on samples that select `CONFIG_PM` explicitly. I could place an `#ifdef CONFIG_TEST` guard around the code that puts the core into wait mode, and that would effectively disable the part of the low power mode code that breaks the debugger whenever a zephyr test is built. Do you think that would work for you? Alternatively if you built tests with `-DCONFIG_PM=n` the power management code should never come into play.

Thanks again for looking over the low power mode PRs, and please just let me know which solution you'd prefer. I definitely don't want to impact CI with the power mode changes.

hakehuang commented:
> @hakehuang `CONFIG_PM` should still not be selected by default, so the debugger should only have issues on samples that select `CONFIG_PM` explicitly. I could place an `#ifdef CONFIG_TEST` guard around the code that puts the core into wait mode, and that would effectively disable the part of the low power mode code that breaks the debugger whenever a zephyr test is built. Do you think that would work for you? Alternatively if you built tests with `-DCONFIG_PM=n` the power management code should never come into play.
> 
> Thanks again for looking over the low power mode PRs, and please just let me know which solution you'd prefer. I definitely don't want to impact CI with the power mode changes.

@danieldegrasse , now even hello world will enter low pow by default, can you check? I prefer that we have a application to demo allow system into lowpower, and by default is it not allowed.

danieldegrasse commented:
@hakehuang When I build hello world for the RT1064, `CONFIG_PM` is not selected. Are you seeing something different? If not, the issue may be from the clocking changes, since if `CONFIG_PM` is not selected, the low power mode entry function `pm_power_state_set` will not be called (see docs [here](https://docs.zephyrproject.org/latest/guides/pm/system.html))

hakehuang commented:
@danieldegrasse , let me update the hal and try again.

danieldegrasse commented:
@hakehuang There was something wrong with the way that I was integrating @mmahadevan108's PR (#41048). I've rebased this PR on top of his, since it depends on it anyways. Can you retry the USB test now? It works for me.

hakehuang commented:
> @hakehuang There was something wrong with the way that I was integrating @mmahadevan108's PR (#41048). I've rebased this PR on top of his, since it depends on it anyways. Can you retry the USB test now? It works for me.

retesting, will feedback once ready

hakehuang commented:
@danieldegrasse , CI is running for this PR, will feedback tomorrow

hakehuang commented:
@danieldegrasse , do you see this failure
```
FAILED: zephyr/zephyr_pre0.elf zephyr/zephyr_pre0.map /home/shared/disk/zephyr_project/zephyr_test/zephyr/twister-out/mimxrt1064_evk/tests/lib/devicetree/devices/libraries.devicetree/zephyr/zephyr_pre0.map
: && ccache /home/ubuntu/zephyr-sdk/arm-zephyr-eabi/bin/arm-zephyr-eabi-gcc   zephyr/CMakeFiles/zephyr_pre0.dir/misc/empty_file.c.obj -o zephyr/zephyr_pre0.elf  zephyr/CMakeFiles/offsets.dir/./arch/arm/core/offsets/offsets.c.obj  -fuse-ld=bfd  -Wl,-T  zephyr/linker_zephyr_pre0.cmd  -Wl,-Map=/home/shared/disk/zephyr_project/zephyr_test/zephyr/twister-out/mimxrt1064_evk/tests/lib/devicetree/devices/libraries.devicetree/zephyr/zephyr_pre0.map  -Wl,--whole-archive  app/libapp.a  zephyr/libzephyr.a  zephyr/arch/common/libarch__common.a  zephyr/arch/arch/arm/core/aarch32/libarch__arm__core__aarch32.a  zephyr/arch/arch/arm/core/aarch32/cortex_m/libarch__arm__core__aarch32__cortex_m.a  zephyr/arch/arch/arm/core/aarch32/mpu/libarch__arm__core__aarch32__mpu.a  zephyr/lib/libc/minimal/liblib__libc__minimal.a  zephyr/lib/posix/liblib__posix.a  zephyr/soc/arm/common/cortex_m/libsoc__arm__common__cortex_m.a  zephyr/boards/arm/mimxrt1064_evk/libboards__arm__mimxrt1064_evk.a  zephyr/subsys/testsuite/ztest/libsubsys__testsuite__ztest.a  zephyr/drivers/clock_control/libdrivers__clock_control.a  zephyr/drivers/console/libdrivers__console.a  zephyr/drivers/gpio/libdrivers__gpio.a  zephyr/drivers/serial/libdrivers__serial.a  zephyr/drivers/timer/libdrivers__timer.a  modules/hal_nxp/hal_nxp/lib..__modules__hal__nxp.a  -Wl,--no-whole-archive  zephyr/kernel/libkernel.a  -L"/home/ubuntu/zephyr-sdk/arm-zephyr-eabi/bin/../lib/gcc/arm-zephyr-eabi/10.3.0/thumb/v7e-m/nofp"  -L/home/shared/disk/zephyr_project/zephyr_test/zephyr/twister-out/mimxrt1064_evk/tests/lib/devicetree/devices/libraries.devicetree/zephyr  -lgcc  zephyr/arch/common/libisr_tables.a  -no-pie  -mcpu=cortex-m7  -mthumb  -mabi=aapcs  -mfp16-format=ieee  -Wl,--gc-sections  -Wl,--build-id=none  -Wl,--sort-common=descending  -Wl,--sort-section=alignment  -Wl,-u,_OffsetAbsSyms  -Wl,-u,_ConfigAbsSyms  -nostdlib  -static  -Wl,-X  -Wl,-N  -Wl,--orphan-handling=warn  -Wl,--fatal-warnings && cd /home/shared/disk/zephyr_project/zephyr_test/zephyr/twister-out/mimxrt1064_evk/tests/lib/devicetree/devices/libraries.devicetree/zephyr && /home/ubuntu/.local/lib/python3.8/site-packages/cmake/data/bin/cmake -E echo
/home/ubuntu/zephyr-sdk/arm-zephyr-eabi/bin/../lib/gcc/arm-zephyr-eabi/10.3.0/../../../../arm-zephyr-eabi/bin/ld.bfd: zephyr/drivers/gpio/libdrivers__gpio.a(gpio_test.c.obj):/home/shared/disk/zephyr_project/zephyr_test/zephyr/drivers/gpio/gpio_test.c:98: multiple definition of `__device_dts_ord_121'; app/libapp.a(main.c.obj):/home/shared/disk/zephyr_project/zephyr_test/zephyr/tests/lib/devicetree/devices/src/main.c:35: first defined here
collect2: error: ld returned 1 exit status
ninja: build stopped: subcommand failed.

```

danieldegrasse commented:
@hakehuang Those tests are passing for me in the latest push (as well as the build failure being fixed) can you verify they are passing for you as well? Thanks for running CI.

hakehuang commented:
@danieldegrasse CI is running, will feedback

danieldegrasse commented:
@hakehuang CI failure should be fixed. Can you verify on your end?

hakehuang commented:
@danieldegrasse i rerun all testing on this board, will feedback tomorrow.

hakehuang commented:
@danieldegrasse still meet issue when run time_api testing.
below is my command line
```
scripts/twister -p mimxrt1064_evk  --device-testing  --west-runner=pyocd --west-flash="--board-id=023200000708c4a700000000000000000000000097969905,--erase"  --device-serial /dev/ttyACM4  -T tests/kernel/timer/timer_api
```
you can update your pyocd board-id and serial port to testing, below is my log

```
INFO    - Adding tasks to the queue...
INFO    - Added initial list of jobs to queue
INFO    - 1/2 mimxrt1064_evk            tests/kernel/timer/timer_api/kernel.timer          PASSED (device 24.096s)
INFO    - 2/2 mimxrt1064_evk            tests/kernel/timer/timer_api/kernel.timer.tickless FAILED Timeout (device 80.326s)
INFO    - /home/shared/disk/zephyr_project/zephyr_test/zephyr/twister-out/mimxrt1064_evk/tests/kernel/timer/timer_api/kernel.timer.tickless/handler.log
ERROR   - *** Booting Zephyr OS build v2.7.99-2905-g34be5e7e81cb  ***
Running test suite timer_api
===================================================================
START - test_time_conversions
PASS - test_time_conversions in 0.8 seconds
===================================================================
START - test_timer_duration_period
PASS - test_timer_duration_period in 0.326 seconds
===================================================================
START - test_timer_restart
PASS - test_timer_restart in 0.276 seconds
===================================================================
START - test_timer_period_0
PASS - test_timer_period_0 in 0.202 seconds
===================================================================
START - test_timer_period_k_forever
PASS - test_timer_period_k_forever in 0.202 seconds
===================================================================
START - test_timer_expirefn_null
PASS - test_timer_expirefn_null in 0.326 seconds
===================================================================
START - test_timer_periodicity
PASS - test_timer_periodicity in 0.252 seconds
===================================================================
START - test_timer_status_get
PASS - test_timer_status_get in 0.1 seconds
===================================================================
START - test_timer_status_get_anytime
PASS - test_timer_status_get_anytime in 0.326 seconds
===================================================================
START - test_timer_status_sync
PASS - test_timer_status_sync in 0.352 seconds
===================================================================
START - test_timer_k_define
PASS - test_timer_k_define in 0.701 seconds
===================================================================
START - test_timer_user_data

INFO    - /home/shared/disk/zephyr_project/zephyr_test/zephyr/twister-out/mimxrt1064_evk/tests/kernel/timer/timer_api/kernel.timer.tickless/handler.log

INFO    - 2 Iteration:
INFO    - Adding tasks to the queue...
INFO    - Added initial list of jobs to queue
INFO    - 3/2 mimxrt1064_evk            tests/kernel/timer/timer_api/kernel.timer.tickless FAILED Flash error (device 1.617s)

INFO    - 3 Iteration:
INFO    - Adding tasks to the queue...
INFO    - Added initial list of jobs to queue
INFO    - 3/2 mimxrt1064_evk            tests/kernel/timer/timer_api/kernel.timer.tickless FAILED Flash error (device 0.840s)
```

danieldegrasse commented:
@hakehuang Should be fixed now- The low power mode disables the UART, and the board was entering low power mode before it could log the results of the final test. Added a check in the SOC level low power mode code to ensure the console transmission is complete before entering LPM

hakehuang commented:
> Should be fixed now- The low power mode disables the UART, and the board was entering low power mode before it could log the results of the final test. Added a check in the SOC level low power mode code to ensure the console transmission is complete before entering LPM

@daniel-thompson , this still has a problem, once the chip enters LPM, the download will fail, so we should not keep in LPM mode, once the test is done.

daniel-thompson commented:
> @daniel-thompson , this still has a problem, once the chip enters LPM, ...

Wrong Daniel?

danieldegrasse commented:
@hakehuang I agree that the chip not flashing while testing PM code is a big issue- but I've had trouble duplicating it. I can add a feature like you describe, but if possible I'd like to duplicate the issue so that I can identify what part of the LPM code is causing flashing to fail, and selectively disable that instead of waking up the device periodically.

To clarify, your setup is the following, right?
- Default openSDA binary flashed to lpc debugger on the RT1064 EVK
- Flashing done by pyocd

Is the bug occurring at random after several twister runs? What would I need to run to reproduce it?

hakehuang commented:
> flashing to fail, and selectively disable that instead of waking up the device periodically.
> 
> To clarify, your setup is the following, right?
> 
> * Default openSDA binary flashed to lpc debugger on the RT1064 EVK
> * Flashing done by pyocd
> 
> Is the bug occurring at random after several twister runs? What would I need to run to reproduce it?

assume your serial port is /dev/ttyACM0 and your pyocd debuger id is 023200000708c4a700000000000000000000000097969905 (`pyocd list` can use to find the id)
```
scripts/twister -p mimxrt1064_evk  --device-testing  --west-runner=pyocd --west-flash="--board-id=023200000708c4a700000000000000000000000097969905,--erase"  --device-serial /dev/ttyACM0  -T tests/ --inline-logs
```

danieldegrasse committed:
tests: ztest: add CONFIG_ZTEST_NO_YIELD to keep SOC awake after test

Some SOCs cannot be flashed reliably in low power modes. If
CONFIG_ZTEST_NO_YIELD is selected, do not yield to the idle thread after
testsuite completes, so that the SOC will not enter low power mode.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse commented:
@hakehuang I've updated this PR to remove the dependency on #41048, and added a block to the ztest subsystem that prevents the SOC from entering low power mode when the test suite completes. Can you run CI again and see if flashing still encounters issues?

hakehuang commented:
> @hakehuang I've updated this PR to remove the dependency on #41048, and added a block to the ztest subsystem that prevents the SOC from entering low power mode when the test suite completes. Can you run CI again and see if flashing still encounters issues?

@daniel-thompson  thanks a lot. CI is running

danieldegrasse commented:
@hakehuang I'm still unable to reproduce the issue on my end unfortunately. I did try running twister against the RT1064 using this PR and saw failures caused by pyocd failing to flash the board, but those same failures appeared when testing with the main branch. I tried flashing the `timer_api` test roughly ten times in a row, and never had any issues. If you have a jlink debugger, could you try running the test suite using it? Or is your test setup not configured to do that easily

mmahadevan108 commented on boards/arm/mimxrt1064_evk/lpm.c:
```diff
@@ -0,0 +1,303 @@
+	while ((CCM_ANALOG->PLL_ENET & CCM_ANALOG_PLL_ENET_LOCK_MASK) == 0) {
+	}
+	CCM_ANALOG->PLL_ENET_CLR = CCM_ANALOG_PLL_ENET_BYPASS_MASK;
+
```
We should probably check if the ETH, Video and Audio PLL's are used before enabling them. For e.g. `CONFIG_ETH_MCUX` could be used to check if ETH PLL should be enabled.

danieldegrasse replied:
I've updated the PR to do this in the latest push. I also added default PLL configurations in the `soc_rt10xx.c` file for the USB2 and AUDIO plls, based off the values chosen by the config tool

mmahadevan108 replied:
Please do not use `INIT_USB_PLL `and `INIT_SYS_PLL `as these have been deleted. You could also avoid using `INIT_ARM_PLL`

danieldegrasse replied:
I've dropped `INIT_USB_PLL` and `INIT_SYS_PLL` in the latest push. I kept `INIT_ARM_PLL` for future compatibility with the RT1010-1024 parts, if we elect to enable this clock management code on those parts.

mmahadevan108 commented on soc/arm/nxp_imx/rt/CMakeLists.txt:
```diff
@@ -25,6 +25,16 @@ if(CONFIG_PM)
+
+zephyr_compile_definitions(
+  XIP_EXTERNAL_FLASH
+  SKIP_SYSCLK_INIT
```
`SKIP_SYSCLK_INIT` can be deleted as this is not used.

danieldegrasse replied:
Removed, thanks

mmahadevan108 commented on boards/arm/mimxrt1064_evk/lpm.c:
```diff
@@ -0,0 +1,312 @@
+/*
```
Is there a reason this file cannot be in the `soc` folder?

danieldegrasse replied:
No there is not. I have moved it to the `soc` folder in the latest push and made the clock code read all PLL configurations at boot, so that changes to PLL dividers on custom boards can be accounted for.

hakehuang commented:
> I'm still unable to reproduce the issue on my end unfortunately. I did try running twister against the RT1064 using this PR and saw failures caused by pyocd failing to flash the board, but those same failures appeared when testing with the main branch. I tried flashing the `timer_api` test roughly ten times in a row, and never had any issues. If you have a jlink debugger, could you try running the test suite using it? Or is your test setup not configured to do that easily

@danieldegrasse ，
···
I find a reproduce steps, flash the time_api with tickless configure, as below

cd /tests/kernel/timer/timer_api

west build -b mimxrt1064_evk -- -DCONF_FILE=prj_tickless.conf

west flash --runner=pyocd

after the test case pass, you can wait for a while and try to run flash again (some times you need try 2-3 times)

west flash --runner=pyocd

it will report failure
···

this issue maybe less reproducable in jlink, as it may has less reset delay, but there are potential in the implementation

danieldegrasse committed:
timers: mcux_gpt_timer: Change MCUX GPT timer to use indirect ISR

Indirect ISR automatically calls power management functions, which GPT
timer direct ISR was not calling. Calling these functions means that the
kernel will recognize that it is exiting low power mode when the GPT
timer interrupt fires that wakes the SOC up, and will call
pm_power_state_exit_post_ops, which can in turn raise the clock
frequencies and voltage of the SOC as early as possible.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
boards: mimxrt1064: Document debugger issue when entering LPM

Add documentation warning user that entering low power modes can cause
debugger to fail to connect.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
soc: mimxrt1064: Enable low power clocking for mimxrt1064

Enables low power clocks for mimxrt1064. This allows the soc to
transition into very low power states during idle, as most PLLs can be
disabled. Power states need to be enabled and selected at the board
level.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
soc: nxp_imx: Enable inclusion of power management HAL drivers

Add KConfig symbols to select power management HAL drivers required for low
power modes. Set ZTEST_NO_YEILD if PM is enabled, since RT series SOCs
do not connect to a debugger in WAIT mode.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse commented:
@hakehuang I've updated this PR with the fixes required to pass CI on the RT1064. Could you run tests/kernel/timer/timer_api on your board one last time to verify it passes?

danieldegrasse commented:
@nashif I made changes to the ztest module in this PR, to explicitly block boards from entering the idle thread when tests complete. On the RT1064, this is required because low power modes will power down the ARM core, and a debugger will not be able to flash new code, so tests that enable `CONFIG_PM` cannot be run using CI on the board. Can you confirm this won't cause issues with testing on other boards?

hakehuang commented:
> @hakehuang I've updated this PR with the fixes required to pass CI on the RT1064. Could you run tests/kernel/timer/timer_api on your board using this PR one last time to verify it passes?

CI is done, now it is OK. Nick work!

tejlmand commented on soc/arm/nxp_imx/rt/CMakeLists.txt:
```diff
@@ -25,6 +25,19 @@ if(CONFIG_PM)
 endif()
 
+if (CONFIG_PM AND CONFIG_SOC_SERIES_IMX_RT10XX)
+zephyr_sources(power_rt10xx.c)
+zephyr_code_relocate(power_rt10xx.c ITCM_TEXT)
```
Please indent for readability:
```suggestion
  zephyr_sources(power_rt10xx.c)
  zephyr_code_relocate(power_rt10xx.c ITCM_TEXT)
```

danieldegrasse replied:
Fixed in latest push

tejlmand commented on soc/arm/nxp_imx/rt/CMakeLists.txt:
```diff
@@ -25,6 +25,19 @@ if(CONFIG_PM)
+zephyr_sources(power_rt10xx.c)
+zephyr_code_relocate(power_rt10xx.c ITCM_TEXT)
+if (CONFIG_SOC_MIMXRT1064)
+zephyr_sources(lpm_rt1064.c)
+zephyr_code_relocate(lpm_rt1064.c ITCM_TEXT)
```
Please indent for readability:
```
  zephyr_sources(lpm_rt1064.c)
  zephyr_code_relocate(lpm_rt1064.c ITCM_TEXT)
```

danieldegrasse replied:
Fixed in latest push

mbolivar-nordic commented:
@gmarull @ceolin could you please ACK?

danieldegrasse commented:
@carlescufi or @nashif, could you look at the changes to `ztest.c`? Our RT parts will not always reliably connect to a debugger when `CONFIG_PM=y` and the kernel requests the SOC enter a sleep state. This impacts CI testing, so I changed `ztest.c` to not exit `main` after tests complete, but instead to lock IRQs and spin, so that the part cannot enter a sleep state. Just want to ensure this won't break any existing tests.

hakehuang commented on subsys/testsuite/ztest/src/ztest.c:
```diff
@@ -562,5 +562,17 @@ void main(void)
 			state.boots = 0;
 		}
 	}
+#ifdef CONFIG_PM
```
for example for mimxrt1064 baord, if board enters low power mode, we can not connect to it with jtag interface.

danieldegrasse replied:
Any test running on an RT1064 will be affected when `CONFIG_PM=y`, and I believe in some tests @hakehuang has ran he has seen issues on the RT1170. The core issue is that when the kernel has no task left to run (main exits) the chip will enter the lowest power idle state, and on the RT parts a debugger cannot reliability connect to the part in that state.

Testing on `native_posix`, tests selecting `CONFIG_PM=y` still pass. The call to `posix_exit` that exists the test is made before the `CONFIG_PM` block at the end of `main`, so posix tests never reach this loop.

danieldegrasse commented on soc/arm/nxp_imx/rt/lpm_rt1064.c:
```diff
@@ -0,0 +1,458 @@
+uint32_t flexspi_div;
+
+/* Internal version of this function used, since we relocate this code to ITCM */
+static inline void CLOCK_SETMUX(clock_mux_t mux, uint32_t value)
+{
+	uint32_t busyShift;
+
+	busyShift = (uint32_t)CCM_TUPLE_BUSY_SHIFT(mux);
+	CCM_TUPLE_REG(CCM, mux) = (CCM_TUPLE_REG(CCM, mux) & (~CCM_TUPLE_MASK(mux))) |
+		  (((uint32_t)((value) << CCM_TUPLE_SHIFT(mux))) & CCM_TUPLE_MASK(mux));
+
+	assert(busyShift <= CCM_NO_BUSY_WAIT);
+
+	/* Clock switch need Handshake? */
+	if (busyShift != CCM_NO_BUSY_WAIT) {
+		/* Wait until CCM internal handshake finish. */
+		while ((CCM->CDHIPR & ((1UL << busyShift))) != 0UL) {
+		}
+	}
+}
```
I'll update this function to use correct coding guidelines. However, I believe it should remain in this file, because this function needs to be linked into ITCM (since it is invoked to configure the clocks of the flash memory and SRAM).

The function could be moved to another file that is also relocated to ITCM, but I would argue it is best to keep all code touching the flash and SRAM clocks in one place. Bugs involving reading from the flash while changing its clocks often are very hard to narrow down, and adding something like a `LOG_DBG` call in this function would be enough to cause issues. Keeping all flash clocking code in one file that is moved to ITCM reduces the risk of a user unintentionally adding function calls here that could introduce bugs.

danieldegrasse committed:
drivers: serial: Add pm constraint setting to mcux lpuart driver

System entering sleep state before uart tx is complete can result in
characters being dropped from the transmission. Add pm constraint
setting to the lpuart driver to prevent character drops

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse commented on soc/arm/nxp_imx/rt/Kconfig.defconfig.series:
```diff
@@ -100,6 +100,9 @@ config PM_MCUX_GPC
 if SOC_SERIES_IMX_RT10XX && PM
 
+config CODE_DATA_RELOCATION
+	default y
```
the conditional on line 101 should make that KConfig equivalent to the following, right?
```
config CODE_DATA_RELOCATION
    default y if SOC_SERIES_IMX_RT10XX && PM
```

danieldegrasse commented on soc/arm/nxp_imx/rt/power_rt10xx.c:
```diff
@@ -0,0 +1,270 @@
+	while (!(kLPUART_TransmissionCompleteFlag &
+		LPUART_GetStatusFlags((LPUART_Type *)DT_REG_ADDR(DT_CHOSEN(zephyr_console))))) {
+		;
+	}
```
Done in the latest push, thank you for letting me know about that functionality in the PM subsystem.

hakehuang reviewed (dismissed):
regression test ALL PASS on rt1064 platform

gmarull commented on soc/arm/nxp_imx/rt/Kconfig.defconfig.series:
```diff
@@ -99,6 +98,23 @@ config PM_MCUX_GPC
+config PM_MCUX_PMU
+	default y if HAS_MCUX_PMU
+
+# Don't allow SOC to sleep after tests complete when PM is enabled
+config ZTEST
+	select ZTEST_NO_YEILD
```
shouldn't this be:

```
config ZTEST_NO_YIELD
    default y if ZTEST
```

danieldegrasse replied:
I chose select because that symbol needs to be enabled unconditionally for RT series SOCs when building with `CONFIG_PM=y and CONFIG_ZTEST=y` .In my view, the build should fail if `CONFIG_ZTEST=y`, `CONFIG_PM=y` and `CONFIG_ZTEST_NO_YEILD=n`, because otherwise twister would proceed to flashing the build, and then the board will effectively be bricked until whoever is running the test recovers it, and the test run would essentially be invalid.

danieldegrasse replied:
After discussing with @hakehuang offline, he said that he would prefer the `default y if ZTEST` approach. Since I made this change with his CI in mind, I've switched the KConfig definition to use `default y if (ZTEST && PM)`. I've also expanded this definition to all RT series parts, not just RT10xx, since twister has issues flashing code to any RT part in low power mode.

gmarull commented on subsys/testsuite/ztest/Kconfig:
```diff
@@ -79,6 +79,14 @@ config ZTEST_ASSERT_HOOK
 	  error test case. Remember to add ignore_fault tag in yaml file when
 	  using twister to run testing.
 
+config ZTEST_NO_YEILD
```
nit: typo: `YIELD`

hakehuang commented on subsys/testsuite/ztest/Kconfig:
```diff
@@ -79,6 +79,14 @@ config ZTEST_ASSERT_HOOK
 	  error test case. Remember to add ignore_fault tag in yaml file when
 	  using twister to run testing.
 
+config ZTEST_NO_YIELD
```
this shall be default y, we can not afford to explore all test  applications to add them platform by platform.

gmarull reviewed (approved):
+1. Note that I still think clock code now in `soc` should likely be generalized or improved. @erwango is now doing great work on the clocks area, see e.g. https://github.com/zephyrproject-rtos/zephyr/pull/42097

ceolin commented on drivers/timer/mcux_gpt_timer.c:
```diff
@@ -133,14 +133,11 @@ static uint32_t elapsed(void)
  *
- * Note: we use a direct ISR for latency concerns.
  */
-ISR_DIRECT_DECLARE(mcux_imx_gpt_isr)
```
s2

ceolin commented on soc/arm/nxp_imx/rt/power_rt10xx.c:
```diff
@@ -0,0 +1,260 @@
+		lpm_enter_sleep_mode(kCLOCK_ModeWait);
+		break;
+	case PM_STATE_SUSPEND_TO_IDLE:
+		LOG_DBG("entering PM state suspend to idle");
```
Not a exclusive problem of this patch, but logging at this point *may* be problematic. Device were already suspended here.

carlescufi commented:
@dleach02 ping
